### ISSUE TITLE
bnd-baseline-maven-plugin doesn't baseline against version ranges

### ISSUE THREAD

**Issue by anselmov** [bndtools/bnd#2315]
As an eclipse bndtools user, I expect the bnd-baseline-maven-plugin to work similarly and baseline against the highest version (latest).

My test is based on [valid-no-previous integration-test](https://github.com/bndtools/bnd/tree/master/maven/bnd-baseline-maven-plugin/src/test/resources/integration-test/test/valid-no-previous). I did the following steps:

1. change the project version to **1.0.0,** run **mvn install** (to local repository), build success
2. same code, change the project version to **3.0.0**, run **mvn install**, build success
3. same code, change the project version to **2.0.0**, run **mvn install**, build success

The step 3 should fail but it baselines against **1.0.0** because the default base search range is: 

> (,${project.version})

I want to baseline against the highest version, so when I override the range in the plugin configuration:
```
<configuration>
    <base>
          <groupId>${project.groupId}</groupId>
          <artifactId>${project.artifactId}</artifactId>
          <version>(${project.version},)</version>
    </base>
</configuration>
```
I clean the .m2 repository, redo steps 1 and 2, then **mvn verify** and I get 

`Failed to execute goal biz.aQute.bnd:bnd-baseline-maven-plugin:4.0.0-SNAPSHOT:baseline (baseline) on project test: Unable to locate a previous version of the artifact: Could not find artifact org.example:test:jar:(2.0.0,) in central (file:///Users/anselmov/.m2/repository)`

I'm not sure if my expectations are too high, but version ranges seem not handled properly. Any help is appreciated.

mvn version: 3.3.3
java version: 1.8.0_152-ea
OSX 10.9.5

**Comment by rotty3000** [comment]
doesn't `(2.0.0,)` mean `> 2.0.0 & < 3.0.0` (i.e. `2.0.x -> 2.x.x`) which does not match `3.0.0`?

**Comment by anselmov** [comment]
`(2.0.0,)` means `> 2.0.0`, I did some tests with:
https://github.com/bndtools/bnd/blob/f553787d0e339948b01085545e291f63d0e9175c/maven/bnd-baseline-maven-plugin/src/main/java/aQute/bnd/maven/baseline/plugin/BaselineMojo.java#L184
There is a workaround when the default range is "any version" like:
`Artifact toCheck = toFind.setVersion("[0,)");`

btw: do you know why this single change makes integration-tests fail? 

> Execution integration-test of goal org.apache.maven.plugins:maven-invoker-plugin:3.0.1:run failed: character 27 is not allowed in output

**Comment by bjhargrave** [comment]
> doesn't (2.0.0,) mean > 2.0.0 & < 3.0.0 (i.e. 2.0.x -> 2.x.x) which does not match 3.0.0?

See http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html. `(x,)` means versions > x.

But using `(2,)` as the version range for the baseline jar seems wrong. You want to baseline against the highest version LESS THAN the version you are currently building. This is why `(,${project.version})` is used. There is an issue (#2248) which complains that SNAPSHOT can be used for the baseline. This needs to be fixed since SNAPSHOTs are, by definition, not released versions.

**Comment by bjhargrave** [comment]
But I do see that the plugin does not correctly handle using versions in the configuration. I am looking at a fix now.

**Comment by anselmov** [comment]
> You want to baseline against the highest version LESS THAN the version you are currently building.

In bndtools, the default behavior is to baseline against **the latest version in the repository**, so we can't release a version between two existing versions. Can you confirm that this difference is intentional?

**Comment by bjhargrave** [comment]
> In bndtools, the default behavior is to baseline against the latest version in the repository, so we can't release a version between two existing versions. Can you confirm that this difference is intentional?

I don't know what Bndtools does here. Sorry.

**Comment by bjhargrave** [comment]
With the fix in #2316, you can specify a range in the configuration and the baseline will use the highest version in the range.

**Comment by timothyjward** [comment]
Just to confirm what others are saying:

> I want to baseline against the highest version, so when I override the range in the plugin configuration:

    <configuration>
        <base>
              <groupId>${project.groupId}</groupId>
              <artifactId>${project.artifactId}</artifactId>
              <version>(${project.version},)</version>
        </base>
     </configuration>

This is a *really* bad way to set up your build. Maven builds are supposed to be repeatable in the future so you should never attempt to baseline against a version that is higher than the version being built. Imagine, for example, that you release version 1.0.0, then later you release version 1.1.0. At this point you can no longer rebuild the tagged version 1.0.0 (or any 1.0.x fixes) because it will fail baselining against 1.1.0.
